src/utils/i18n.ts

```diff
@@ -0,0 +1,264 @@
+import { PluralCategory, other } from './pluralization';
+import LanguageTag from './LanguageTag';
+
+/**
+ * Utility methods and interfaces that are used by Mnoga and can be used in applications that want
+ * to handle their own localization state.
+ */
+
+// Define some variables for validating keys and key context.
+const CONTEXT_CHARACTERS = '[A-Za-z0-9-_]+';
+const DELIMITER = '.';
+
+// Used to find variable names inside the output string.
+const INTERPOLATION_REGEX = /%{([^}]*)}/g;
+
+export const VALID_KEY_CONTEXT = new RegExp(`^${CONTEXT_CHARACTERS}$`);
+const VALID_KEY = new RegExp(`^${CONTEXT_CHARACTERS}(${DELIMITER}${CONTEXT_CHARACTERS})*$`);
+
+/**
+ * Aliases are a way of saying that one locale should be substituted for another locale. These
+ * should be used when a locale is not officially supported, but is preferred to fallback on a
+ * different locale. For example, `en-AU` could point to `en-GB` if we wanted Australians to see
+ * British English translations, instead of falling back to `en`.
+ */
+export interface Aliases {
+ [propName: string]: string | undefined;
+}
+
+interface Set {
+  [propName: string]: boolean | undefined;
+}
+
+export interface Translations {
+  [propName: string]: Translations | string | undefined;
+}
+
+export interface TData {
+  count?: number;
+  [propName: string]: string | number | undefined;
+}
+
+/**
+ * @param data
+ * @param delimiter
+ * @param interpolationFormat
+ * @param key
+ * @param keyFormat
+ * @param locales
+ * @param rules
+ * @param translations
+ */
+export interface TOptions {
+  data: TData;
+  delimiter?: string;
+  interpolationFormat?: RegExp;
+  key: string;
+  keyFormat?: RegExp;
+  locales: string[];
+  rules?: Rules;
+  translations: Translations;
+}
+
+export interface Rule {
+  (n: number): PluralCategory;
+}
+
+export interface Rules {
+  [propName: string]: Rule | undefined;
+}
+
+/**
+ * Performs a locale lookup for a single best locale. This performs quite similarly to
+ * RFC documentation on looking up locales: https://tools.ietf.org/html/rfc4647#section-3.4.
```

Comment:
There's actually a RFC documentation describing how this should be done. The behavior is mostly the same, although we support substitutes.